src/github.com/matrix-org/dendrite/roomserver/input/state.go

```diff
@@ -0,0 +1,258 @@
+		// Check that those state events are in the database and store the state.
+		entries, err := db.StateEntriesForEventIDs(stateEventIDs)
+		if err != nil {
+			return 0, err
+		}
+
+		return db.AddState(roomNID, nil, entries)
+	}
+
+	// Load the state at the prev events.
+	prevEventRefs := event.PrevEvents()
+	prevEventIDs := make([]string, len(prevEventRefs))
+	for i := range prevEventRefs {
+		prevEventIDs[i] = prevEventRefs[i].EventID
+	}
+
+	prevStates, err := db.StateAtEventIDs(prevEventIDs)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(prevStates) == 0 {
+		// 2) There weren't any prev_events for this event so the state is
+		// empty.
+		return db.AddState(roomNID, nil, nil)
+	}
+
+	if len(prevStates) == 1 {
+		prevState := prevStates[0]
+		if prevState.EventStateKeyNID == 0 {
+			// 3) None of the previous events were state events and they all
+			// have the same state, so this event has exactly the same state
+			// as the previous events.
+			// This should be the common case.
+			return prevState.BeforeStateSnapshotNID, nil
+		}
+		// The previous event was a state event so we need to store a copy
+		// of the previous state updated with that event.
+		stateDataNIDLists, err := db.StateDataNIDs([]types.StateSnapshotNID{prevState.BeforeStateSnapshotNID})
+		if err != nil {
+			return 0, err
+		}
+		stateDataNIDs := stateDataNIDLists[0].StateDataNIDs
+		if len(stateDataNIDs) < maxStateDataNIDs {
+			// 4) The number of state data blocks is small enough that we can just
+			// add the state event as a block of size one to the end of the blocks.
+			return db.AddState(
+				roomNID, stateDataNIDs, []types.StateEntry{prevState.StateEntry},
+			)
+		}
+		// If there are too many deltas then we need to calculate the full state
+		// So fall through to calculateAndStoreStateMany
+	}
+	return calculateAndStoreStateMany(db, roomNID, prevStates)
+}
+
+const maxStateDataNIDs = 64
+
+func calculateAndStoreStateMany(db RoomEventDatabase, roomNID types.RoomNID, prevStates []types.StateAtEvent) (types.StateSnapshotNID, error) {
+	// Conflict resolution.
+	// First stage: load the state datablocks for the prev events.
+	stateNIDs := make([]types.StateSnapshotNID, len(prevStates))
+	for i, state := range prevStates {
+		stateNIDs[i] = state.BeforeStateSnapshotNID
+	}
+	stateDataNIDLists, err := db.StateDataNIDs(uniqueStateSnapshotNIDs(stateNIDs))
+	if err != nil {
+		return 0, err
+	}
+
+	var stateDataNIDs []types.StateDataNID
+	for _, list := range stateDataNIDLists {
+		stateDataNIDs = append(stateDataNIDs, list.StateDataNIDs...)
+	}
+	stateEntryLists, err := db.StateEntries(uniqueStateDataNIDs(stateDataNIDs))
+	if err != nil {
+		return 0, err
+	}
+	stateDataNIDsMap := stateDataNIDListMap(stateDataNIDLists)
+	stateEntriesMap := stateEntryListMap(stateEntryLists)
+
+	var combined []types.StateEntry
+	for _, prevState := range prevStates {
+		list, ok := stateDataNIDsMap.lookup(prevState.BeforeStateSnapshotNID)
```

Comment:
RAWR! `list` is possibly the most unhelpful variable name ever. It's a list of WHAT (how many lists does this function deal with, be clearer!).